CesiumGltf/include/CesiumGltf/SharedAssetDepot.h

```diff
@@ -0,0 +1,260 @@
+#pragma once
+
+#include <CesiumAsync/AsyncSystem.h>
+#include <CesiumAsync/Future.h>
+#include <CesiumAsync/IAssetAccessor.h>
+#include <CesiumAsync/IAssetResponse.h>
+#include <CesiumUtility/IntrusivePointer.h>
+#include <CesiumUtility/ReferenceCounted.h>
+
+#include <cstddef>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+namespace CesiumGltf {
+
+template <typename T> class SharedAsset;
+
+/**
+ * A depot for {@link SharedAsset} instances, which are potentially shared between multiple objects.
+ * @tparam AssetType The type of asset stored in this depot. This should usually
+ * be derived from {@link SharedAsset}.
+ */
+template <typename AssetType>
+class SharedAssetDepot : public CesiumUtility::ReferenceCountedThreadSafe<
+                             SharedAssetDepot<AssetType>> {
+public:
+  /**
+   * @brief The maximum total byte usage of assets that have been loaded but are
+   * no longer needed.
+   *
+   * When cached assets are no longer needed, they're marked as
+   * candidates for deletion. However, this deletion doesn't actually occur
+   * until the total byte usage of deletion candidates exceeds this threshold.
+   * At that point, assets are cleaned up in the order that they were marked for
+   * deletion until the total dips below this threshold again.
+   *
+   * Default is 100MB.
+   */
+  int64_t staleAssetSizeLimit = 1000 * 1000 * 100;
```

Comment:
This default seems a little high to me (especially on mobile devices), given that it's memory used by assets that are theoretically shareable between models, but that currently aren't used by any of them. We may have to get more experience with different tilesets to really decide, but in the meantime I think it's better to err on the side of "too small" (which can only force us to load an image from the request cache again) rather than "too large" (which can waste GPU memory in users' applications). The similar "sub-tile cache" in raster overlays defaults to 16 MiB (`16 * 1024 * 1024`), and I think that would be a reasonable default here, too.